### PR TITLE
Update wallet server to support retry example

### PR DESCRIPTION
**Pull request by dapengzhang0**
Change wallet server as described in the retry example http://go/grpc-wallet-example-update-for-retry

> forward the 'route' request header from wallet service to account service

### CODE REVIEW

**Comment by dapengzhang0** [comment]
@srini100 @donnadionne @yashykt @menghanl PTAL

Test passed with Java server and Java and Cpp client.

Go client was not tested because xds retry was not implemented.
Cpp server was not tested.
Unfortunately Go server was not tested because I wasn't able to bring go server HEALTHY, even without any code change and keep recreating the fw rule every 2 minutes.

**Comment by srini100** [comment]
The url_map_template.yaml also needs to be updated, right?

**Comment by dapengzhang0** [comment]
> The url_map_template.yaml also needs to be updated, right?

Yes, the url_map_template.yaml needs to be updated, not only for retry, so done separately, see #60.

File: java/build.gradle [hunk 1]
```diff
@@ -17,7 +17,7 @@ repositories {
 sourceCompatibility = 1.8
 targetCompatibility = 1.8
 
-def grpcVersion = '1.37.0' // CURRENT_GRPC_VERSION
+def grpcVersion = '1.40.0' // CURRENT_GRPC_VERSION
```
Comment (menghanl) [inline java/build.gradle:20]: Also update the version for cpp? Is it available now?

File: java/src/main/java/io/grpc/examples/wallet/WalletServer.java [hunk 2]
```diff
@@ -170,31 +171,19 @@ private void start() throws IOException {
...
-                  ServerInterceptors.intercept(
-                      new WalletImpl(accountChannel, statsChannel, v1Behavior),
-                      new WalletInterceptors.HostnameInterceptor(),
-                      new WalletInterceptors.AuthInterceptor()))
-              .addService(ProtoReflectionService.newInstance())
-              .addService(health.getHealthService())
-              .build()
-              .start();
-    }
+
```
Comment (menghanl) [inline java/src/main/java/io/grpc/examples/wallet/WalletServer.java:174]: Why did you delete this?

**Review by menghanl — CHANGES_REQUESTED** [review]

File: java/src/main/java/io/grpc/examples/wallet/WalletServer.java (hunk 2 above)
Comment (dapengzhang0) [reply java/src/main/java/io/grpc/examples/wallet/WalletServer.java:174]: Can use the same code path for the two `credentialsType`s. Previously `InsecureServerCredentials.create()` was not used.

File: java/build.gradle (hunk 1 above)
Comment (dapengzhang0) [reply java/build.gradle:20]: It's not available yet, but should be coming out soon. Updated, but will not merge until it's available.

File: java/src/main/java/io/grpc/examples/wallet/WalletServer.java (hunk 2 above)
Comment (menghanl) [reply java/src/main/java/io/grpc/examples/wallet/WalletServer.java:174]: Are `XdsServerBuilder` and `ServerBuilder` the same?
Doesn't `XdsServerBuilder` wait for LDS? Or that behavior depends on the creds type?

Comment (dapengzhang0) [reply java/src/main/java/io/grpc/examples/wallet/WalletServer.java:174]: For 1.40 they are the same. From next release, it will wait for LDS. What is the plan? Do we want NOT to apply FilterChain if creds type is "insecure" even in the future releases?

cc @sanjaypujare

Comment (dapengzhang0) [reply java/src/main/java/io/grpc/examples/wallet/WalletServer.java]: I changed it back, not making unrelated change.

**Review by menghanl — APPROVED** [review]

**Review by yashykt — APPROVED** [review]
